src/nvim/buffer.c

```diff
@@ -4127,6 +4128,7 @@ void buf_signcols_del_check(buf_T *buf, linenr_T line1, linenr_T line2)
 void buf_signcols_add_check(buf_T *buf, sign_entry_T *added)
 {
   if (!buf->b_signcols.valid) {
+    invalidate_statuscol(NULL, buf);
```

Comment:
Are you also sure this is the right place to add this? The signcol could be invalidated below and statuscol wouldn't be reset.

Wouldn't it make more sense to invalidate status whenever this function (or the one above) gets called?

Reply:
No I'm not sure, just going by real world testing by adding and removing diagnostic signs in a buffer. Haven't encountered a situation where that early return is not entered.

Reply:
In that case, this should probably be called whenever `b->signcols.valid = false`.

Reply:
So the reason that works is that when `'signcolumn' == "no"`, we early return here:
https://github.com/neovim/neovim/blob/86c164d2a14382adfb388a66699828df6c623102/src/nvim/option.c#L5518-L5523
Which in turn means that `buf_signcols()` never gets called and thus `b->signcols.valid` is never validated, meaning that `buf_signcols_add/del_check()` always return early and thus always invalidate the statuscolumn in this PR.

I guess we could add a `*wp->w_p_stc != NUL` check to make sure that `win_signcol_configured()` doesn't return early so that `b->signcols.valid` is validated.
Then I think we should indeed invalidate whenever `v->signcols.valid` is set to `false`.